tokio-trace/src/value.rs

```diff
@@ -0,0 +1,363 @@
+//!
+//! Typically, we might accept arbitrarily-typed values by making the
+//! `Subscriber` APIs that accept them generic. However, as the `Dispatch` type
+//! holds a subscriber as a boxed trait object, the `Subscriber` trait must be
+//! object-safe --- it cannot have trait methods that accept a generic
+//! parameter. Thus, we erase the value's original type.
+//!
+//! However, a `Value` which is valid for the `'static` lifetime may also be
+//! _downcast_ back to a concrete type, similarly to `std::error::Error` and
+//! `std::any::Any`. If the erased type of the value is known, downcasting
+//! allows it to be used as an instance of that type.
+//!
+//! # `Value`s and `Subscriber`s
+//!
+//! `Subscriber`s consume `Value`s as fields attached to `Event`s or `Span`s.
+//! These cases are handled somewhat differently.
+//!
+//! When a field is attached to an `Event`, the `Subscriber::observe_event`
+//! method is passed an `Event` struct which provides an iterator
+//! (`Event::fields`) to iterate over the event's fields, providing references
+//! to the values as `BorrowedValue`s. Since an `Event` represents a _moment_ in
+//! time, it   does not expect to outlive the scope that created it. Thus, the
+//! values attached to an `Event` are _borrowed_ from the scope where the
+//! `Event` originated.
+//!
+//! `Span`s, on the other hand, are somewhat more complex. A `Span` may outlive
+//! scope in which it was created, and as `Span`s are not instantaneous, the
+//! values of their fields may be discovered and added to the span _during_ the
+//! `Span`'s execution. Thus, rather than receiving all the field values when
+//! the span is initially created, subscribers are instead notified of each
+//! field as it is added to the span, via the `Subscriber::add_value` method.
+//! That method is called with the span's ID, the name of the field whose value
+//! is being added, and the value to add.
+//!
+//! Since spans may have arbitrarily long lifetimes, passing the subscriber a
+//! `&dyn AsValue` isn't sufficient. Instead, if a subscriber wishes to persist a
+//! span value for the entire lifetime of the span, it needs the ability to
+//! convert the value into a form in which it is owned by the _subscriber_,
+//! rather than the scope in which it was added to the span. For this reason,
+//! span values are passed as `&dyn IntoValue`. The `IntoValue` trait is an
+//! extension of the `Value` trait that allows conversion into an `OwnedValue`,
+//! a type which represents an owned value allocated on the heap. Since some
+//! subscriber implementations may _not_ need to persist span field values
+//! indefinitely, they are not heap-allocated by default, to avoid unnecessary
+//! allocations, but the `IntoValue` trait presents `Subscriber`s with the
+//! _option_ to box values should they need to do so.
+use std::{any::TypeId, borrow::Borrow, fmt};
+
+/// A formattable field value of an erased type.
+pub trait Value: fmt::Debug + Send + Sync {
```

Comment:
Bikeshed: this trait uses `fmt::Debug` to represent the "canonical" way of formatting a value (whatever the wrapped `AsValue` type defines as its `fmt_value` method. Perhaps this should be `fmt::Display` instead? Internally it _may_ be either, as determined at the call site.